src/projects/EnsureThat/Enforcers/StringArg.cs

```diff
@@ -270,6 +271,19 @@ public string IsInRange(string value, string min, string max, StringComparison c
             return value;
         }
 
+        public string IsAllLettersOrDigits([ValidatedNotNull] string value, [InvokerParameterName] string paramName = null, OptsFn optsFn = null)
+        {
+            Ensure.Any.IsNotNull(value, paramName, optsFn);
+
+            if (!value.ToCharArray().All(char.IsLetterOrDigit))
```

Comment:
If we only had targeted .NETStandard2+ `ToCharArray` would have been redundant. Thinking that we perhaps should use a `for`  construct instead of preprocessor directive. E.g.

```csharp
for(var i = 0; i < value.Length; i++)
    if(!char.IsLetterOrDigit(value[i]))
        throw Ensure.ExceptionFactory.ArgumentException(
            string.Format(ExceptionMessages.Strings_IsAllLettersOrDigits_Failed, value),
            paramName,
            optsFn);
```
@ndrwrbgs opinions on this?

Reply:
Would you like this iterative version instead ?
  
I'm also pretty new to .Net, could you explain me what is the issue here ? I'm a bit curious and would like to understand it a little better if you don't mind enlighten me 😄

Reply:
I was thinking there would be unnecessary allocations. Did a benchmark using BenchmarkDotNet.


```csharp
using System;
using System.Linq;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Order;
using BenchmarkDotNet.Running;

namespace ConsoleApp13
{
    class Program
    {
        static void Main(string[] args) => BenchmarkRunner.Run<MyBenchmarks>();
    }

    [Orderer(SummaryOrderPolicy.FastestToSlowest)]
    [RankColumn]
    [MemoryDiagnoser]
    [CategoriesColumn]
    public class MyBenchmarks
    {
        private static readonly string Value = Guid.NewGuid().ToString("N");

        [Benchmark]
        public void For()
        {
            for (var i = 0; i < Value.Length; i++)
                if (!char.IsLetterOrDigit(Value[i]))
                    throw new Exception("FAIL");
        }

        [Benchmark]
        public void ToCharArray()
        {
            if(!Value.ToCharArray().All(char.IsLetterOrDigit))
                throw new Exception("FAIL");
        }

        [Benchmark]
        public void All()
        {
            if (!Value.All(char.IsLetterOrDigit))
                throw new Exception("FAIL");
        }

        [Benchmark]
        public void All_ExplicitLambda()
        {
            if (!Value.All(c => char.IsLetterOrDigit(c)))
                throw new Exception("FAIL");
        }
    }
}
```

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.836 (1909/November2018Update/19H2)
Intel Xeon CPU E3-1505M v5 2.80GHz, 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.300
  [Host]     : .NET Core 3.1.4 (CoreCLR 4.700.20.20201, CoreFX 4.700.20.22101), X64 RyuJIT
  DefaultJob : .NET Core 3.1.4 (CoreCLR 4.700.20.20201, CoreFX 4.700.20.22101), X64 RyuJIT


```
|             Method |      Mean |    Error |   StdDev | Rank |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------- |----------:|---------:|---------:|-----:|-------:|------:|------:|----------:|
|                For |  68.92 ns | 1.050 ns | 0.820 ns |    1 |      - |     - |     - |         - |
| All_ExplicitLambda | 290.51 ns | 5.664 ns | 8.302 ns |    2 | 0.0076 |     - |     - |      32 B |
|        ToCharArray | 292.90 ns | 4.207 ns | 3.935 ns |    2 | 0.0439 |     - |     - |     184 B |
|                All | 295.69 ns | 5.779 ns | 6.655 ns |    2 | 0.0229 |     - |     - |      96 B |

Reply:
That's really interesting, thank you ! I'm going to adjust it right away